### PR TITLE
fix: Refactoring code to fix a couple of issues related to modules on EE

### PR DESCRIPTION
**Pull request by ankitakinger**
## Description

Refactoring code to fix a couple of issues related to modules on EE

#### PR fixes following issue(s)
Fixes [#29842](https://github.com/appsmithorg/appsmith/issues/29842) [#29445](https://github.com/appsmithorg/appsmith/issues/29445)

#### Type of change
- Bug fix (non-breaking change which fixes an issue)

#### How Has This Been Tested?
- [x] Manual
- [ ] JUnit
- [ ] Jest
- [x] Cypress

#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

## Summary by CodeRabbit

- **New Features**
  - Introduced a new interface for improved action context handling.
  - Added functionality to create API actions based on the editor type.
  - Implemented a new hook for retrieving parent entity information in the datasource editor.

- **Refactor**
  - Updated various components to utilize the new `ActionParentEntityTypeInterface`.
  - Enhanced reducer logic for better handling of action configurations.
  - Streamlined entity exports and imports for consistency.

- **Bug Fixes**
  - Fixed redirection logic in the Datasource Blank State component for a smoother user experience.

- **Chores**
  - Removed unused event listeners from sagas to optimize performance.

### CODE REVIEW

File: app/client/src/ce/entities/URLRedirect/URLAssembly.ts [hunk 1]
```diff
@@ -23,6 +23,7 @@ export interface URLBuilderParams {
   persistExistingParams?: boolean;
   // This is used to pass ID if the sender doesn't know the type of the entity
   parentEntityId?: string;
+  generateEditorPath?: boolean;
```
Comment (ashit-rath) [inline app/client/src/ce/entities/URLRedirect/URLAssembly.ts:26]: What is this used for?

File: app/client/src/pages/Editor/IntegrationEditor/NewApi.tsx [hunk 2]
```diff
@@ -20,6 +18,10 @@ import { getIsGeneratePageInitiator } from "utils/GenerateCrudUtil";
 import { curlImportPageURL } from "@appsmith/RouteBuilder";
 import { getAssetUrl } from "@appsmith/utils/airgapHelpers";
 import { Spinner } from "design-system";
+import { useEditorType } from "@appsmith/hooks";
+import { useParentEntityInfo } from "@appsmith/hooks/datasourceEditorHooks";
+import { createNewApiActionBasedOnEditorType } from "@appsmith/actions/helpers";
+import type { ActionParentEntityTypeInterface } from "@appsmith/entities/Engine/actionHelpers";
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/pages/Editor/IntegrationEditor/NewApi.tsx:24]: tested it.

File: app/client/src/ce/reducers/entityReducers/jsActionsReducer.tsx [hunk 3]
```diff
@@ -117,7 +117,13 @@ export const handlers = {
         return {
           ...a,
           isLoading: false,
-          config: action.payload.data,
+          config: action.payload.data.isPublic
+            ? {
+                ...action.payload.data,
+                isMainJSCollection: true,
+                displayName: "Main",
+              }
```
[the remark this answers is not in the document]
Comment (ankitakinger) [reply app/client/src/ce/reducers/entityReducers/jsActionsReducer.tsx:125]: this is intentional and tested out

File: app/client/src/ce/entities/URLRedirect/URLAssembly.ts (hunk 1 above)
Comment (ankitakinger) [reply app/client/src/ce/entities/URLRedirect/URLAssembly.ts:26]: this is to solve the redirection issue for new datasource button in sub-menu of private entities.

**Review by ashit-rath — APPROVED** [review]

**Pull request merged**